### ISSUE TITLE
Externally deleted emails show up in the inbox and cannot be deleted

### ISSUE THREAD

**Issue by LostinSpacetime** [nextcloud/mail#9168]
### Steps to reproduce

1. Open your email account in Thunderbird with new emails
2. Delete one or more emails before reading them
3. Open the same account in the Mail app on the cloud

### Expected behavior

The inbox shouldn't contain the deleted emails.

### Actual behavior

The inbox contains the deleted emails and the emails are marked as unread. The overall counter for unread emails is 0.
If one clicks on one of the emails, instead of the content on the email, "not found" is shown. Trying to delete the emails results in failure with the message "Message could not be deleted" and after a few seconds the "ghost message" reappears in the inbox.

### Mail app version

3.5.0

### Mailserver or service

IMAP

### Operating system

Ubuntu 18.04

### PHP engine version

PHP 8.1

### Web server

Apache (supported)

### Database

MariaDB

**Comment by LostinSpacetime** [comment]
Currently, my inbox is filled with non-existent emails, and is basically unusable. I would be thankful for a way to rescan my emails as a temporary workaround.

**Comment by 3Nigma** [comment]
Same here

**Comment by alexh-swdev** [comment]
Same here:
* Mail 3.5
* NC 28.0.1
* PHP 8.2.14 (fpm w/ Apache 2.4.46)
* Debian bullseye

I tried the commands w/o success:
* mail:clean-up
* mail:sync

I could not figure out a rule which mails I can remove from my local MUA (which is emClient in my case) but remain in NC-Mail. But there should definetly be an easy way to remophe orphanded mails (mail which are not found) as a hotfix ASAP unless this is fixed properly

**Comment by paulvt** [comment]
It is possible to use "Clear mailbox" on my Inbox, and then the messages are gone.
However, because I also suffer from #5323, when I go to the Priority inbox, they are still all there!
Then, when I mark them as Unimportant and reload the page/app, they are finally all gone.

So the issue seems to be purely that threads and prioritization tags keep tracks of messages that are gone.

**Comment by thrommel1980** [comment]
Same problem here. 
I already avoid deleting emails in the K9 Mail app. If I do, they are in my NC app as corpses. But are deleted in my mail provider inbox.

**Comment by mayermart** [comment]
I do not think this is a Problem caused by K9 itself.
After reading some issues and some experiments i got rid of the issues by simply disable the "Mark E-Mails as Important" Feature in the Nextcloud Mail Settings.

To delete all remaining (Ghost-)E-Mails from the Mailbox i
1. moved all actually existing Mails via Thunderbird (K9 would work the same i guess) in a new temporary Folder
2. used the "Clear mailbox" option on the Inbox folder in Nextcloud Mail and 
3. moved all Mails from the temporary Folder back into my Inbox via Thunderbird.

**Comment by maximelehericy** [comment]
same is happening to me, with archived or sent to junk emails.

**Comment by violoncelloCH** [comment]
posting the full log lines in case that helps:
```
{"reqId":"YFRxUA0D1JXRPoiRvzhL","level":3,"time":"2024-01-22T20:56:55+00:00","remoteAddr":"172.22.0.4","user":"jonas","app":"mail","method":"DELETE","url":"/apps/mail/api/thread/36311","message":"OCA\\Mail\\IMAP\\MessageMapper::move(): Return value must be of type int, null returned in file '/var/www/html/custom_apps/mail/lib/IMAP/MessageMapper.php' line 334","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:121.0) Gecko/20100101 Firefox/121.0","version":"27.1.5.1","exception":{"Exception":"Exception","Message":"OCA\\Mail\\IMAP\\MessageMapper::move(): Return value must be of type int, null returned in file '/var/www/html/custom_apps/mail/lib/IMAP/MessageMapper.php' line 334","Code":0,"Trace":[{"file":"/var/www/html/lib/private/AppFramework/App.php","line":183,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[["OCA\\Mail\\Controller\\ThreadController"],"delete"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":315,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\Mail\\Controller\\ThreadController","delete",["OC\\AppFramework\\DependencyInjection\\DIContainer"],["36311","mail.thread.delete"]]},{"file":"/var/www/html/lib/base.php","line":1068,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/mail/api/thread/36311"]},{"file":"/var/www/html/index.php","line":38,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","Line":169,"Previous":{"Exception":"TypeError","Message":"OCA\\Mail\\IMAP\\MessageMapper::move(): Return value must be of type int, null returned","Code":0,"Trace":[{"file":"/var/www/html/custom_apps/mail/lib/Service/MailManager.php","line":381,"function":"move","class":"OCA\\Mail\\IMAP\\MessageMapper","type":"->","args":[["OCA\\Mail\\IMAP\\ImapClientRateLimitingDecorator",["HICenv","HICflags","HIChdrs","HICdate","HICsize","HICstruct"],true,true],"INBOX",1582,"Trash"]},{"file":"/var/www/html/custom_apps/mail/lib/Service/MailManager.php","line":340,"function":"deleteMessageWithClient","class":"OCA\\Mail\\Service\\MailManager","type":"->","args":[["OCA\\Mail\\Account"],["OCA\\Mail\\Db\\Mailbox",8],1582,["OCA\\Mail\\IMAP\\ImapClientRateLimitingDecorator",["HICenv","HICflags","HIChdrs","HICdate","HICsize","HICstruct"],true,true]]},{"file":"/var/www/html/custom_apps/mail/lib/Service/MailManager.php","line":954,"function":"deleteMessage","class":"OCA\\Mail\\Service\\MailManager","type":"->","args":[["OCA\\Mail\\Account"],"INBOX",1582]},{"file":"/var/www/html/custom_apps/mail/lib/Controller/ThreadController.php","line":205,"function":"deleteThread","class":"OCA\\Mail\\Service\\MailManager","type":"->","args":[["OCA\\Mail\\Account"],["OCA\\Mail\\Db\\Mailbox",8],"<0.0.5.6C3.1DA48ECB54B37CA.0@relationship.automecaniccorbella.com>"]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":230,"function":"delete","class":"OCA\\Mail\\Controller\\ThreadController","type":"->","args":[36311]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":137,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[["OCA\\Mail\\Controller\\ThreadController"],"delete"]},{"file":"/var/www/html/lib/private/AppFramework/App.php","line":183,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[["OCA\\Mail\\Controller\\ThreadController"],"delete"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":315,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\Mail\\Controller\\ThreadController","delete",["OC\\AppFramework\\DependencyInjection\\DIContainer"],["36311","mail.thread.delete"]]},{"file":"/var/www/html/lib/base.php","line":1068,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/mail/api/thread/36311"]},{"file":"/var/www/html/index.php","line":38,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/custom_apps/mail/lib/IMAP/MessageMapper.php","Line":334},"message":"OCA\\Mail\\IMAP\\MessageMapper::move(): Return value must be of type int, null returned in file '/var/www/html/custom_apps/mail/lib/IMAP/MessageMapper.php' line 334","exception":[],"CustomMessage":"OCA\\Mail\\IMAP\\MessageMapper::move(): Return value must be of type int, null returned in file '/var/www/html/custom_apps/mail/lib/IMAP/MessageMapper.php' line 334"},"id":"65aed6b3751ba"}
{"reqId":"YFRxUA0D1JXRPoiRvzhL","level":3,"time":"2024-01-22T20:56:55+00:00","remoteAddr":"172.22.0.4","user":"jonas","app":"PHP","method":"DELETE","url":"/apps/mail/api/thread/36311","message":"Undefined array key 1582 at /var/www/html/custom_apps/mail/lib/IMAP/MessageMapper.php#334","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:121.0) Gecko/20100101 Firefox/121.0","version":"27.1.5.1","data":{"app":"PHP"},"id":"65aed6b3751f5"}
```

**Comment by LostinSpacetime** [comment]
> 
> This is not a fix, but a workaround that usually helps (granted we should not need to run in debug mode to workaround this issue):
> 
>     * Enable Debug mode in Nextcloud `config. php`:
>       
>       * `'debug' => true,`
> 
>     * In the Nextcloud Mail app, click the `...` menu on the folder exhibiting the issue
> 
>     * Choose `Clear cache` (Only available in debug mode)

Thank you! I will try this on the weekend, since I would rather not put the cloud in Debug mode on a working day. We are using Nextcloud in our school and several teachers are complaining about this issue. Is there a possibility to clear the cache for all or certain users?

**Comment by caltenhofen** [comment]
I agree, it would be great to have this as an option for all users.
Or is there any risk in doing so?

**Comment by HLFH** [comment]
This issue is a duplicate of https://github.com/nextcloud/mail/issues/9048.

**Comment by alexh-swdev** [comment]
After skimming the other issue: Ni!

**Comment by caltenhofen** [comment]
I agree from my understanding it's probably a similar but distinct issue. But as I am not too familiar with the codebase the solutions might be connected.

**Comment by violoncelloCH** [comment]
I observed that this particularly happens for me for mails that get classified as "important" by Nextcloud Mail. Not sure if this is a necessary requirement, but for the last few occurrences I could observe it was always mails tagged as important that then stick in the Nextcloud Mail UI. 
Interesting is also, that I can still mark the mail as "not important" while all other actions don't work on the ghost mail.
So maybe that's somehow related to how that "important" mail feature interacts with the mailbox?

**Comment by LostinSpacetime** [comment]
I want to add that even after turning off the priority emails in the app-settings, in Thunderbird deleted emails keep showing up in the app.

**Comment by YourWishes** [comment]
Looking for a solution also. Perhaps related to my use case of running a docker image with both PSQL and redis connected?

**Comment by c0r3dump3d** [comment]
I also have the same problem, in addition to emails it also happens to me with folders, once they are deleted from the imap server from a external client they are not reflected in Nextcloud. NC 28.0.3 and Mail app 3.5.7.

**Comment by mayermart** [comment]
I'd like to correct myself:
Contrary to my previous comment and like @LostinSpacetime already added, disabling Priority E-Mails was no permanent solution.

**Comment by AlexanderZuehr** [comment]
Any news with this problem, a lot of our users having this problem and I am not able to activate debug mode and delete cache for everyone.

**Comment by pedrolucasbp** [comment]
I've tested with Roundcube, K9 and Thunderbird. I confirm the @evilphish comment. All IMAP Clients are OK with deleted messages from other clients, only Nextcloud Mail have this issue. 

Any news for this problem? 

Thank you in advance.

**Comment by blacklight** [comment]
Any progress with this, release plans or workarounds? This issue has been the single blocker preventing me from adopting NC Mail so far. I use a multitude of mail clients, and every time I open NC Mail I'm faced with the "unremovable ghost messages" issue.

**Comment by rikmeijer** [comment]
I created a temporary work-around (am using Nextcloud AIO)

/var/lib/postgresql/data/psql (container)
```
#!/bin/sh

export PGUSER=$POSTGRES_USER
export PGPASSWORD=$POSTGRES_PASSWORD
export PGDATABASE=$POSTGRES_DB

psql "$@"
```
/var/lib/postgresql/data/psql-clear-messages (container)
```
#!/bin/sh

/var/lib/postgresql/data/psql -c "DELETE from oc_mail_recipients WHERE message_id IN (SELECT id FROM oc_mail_messages WHERE mailbox_id=$1);"
/var/lib/postgresql/data/psql -c "DELETE FROM oc_mail_messages WHERE mailbox_id=$1;"
```

On host machine (in .bashrc):
```
alias nc-clear-mail='sudo docker exec -it nextcloud-aio-database /var/lib/postgresql/data/psql-clear-messages'
```

Crontab on host machine
```
@hourly nc-clear-mail <MAILBOX_ID>; occ mail:account:sync <MAIL_ACCOUNT_ID> --force
```

It is nasty, but it keeps my orphaned mails away

**Comment by GVLLIFESTYLE** [comment]
When will this be fixed?

**Comment by wriver4** [comment]
I had a similar issue and went into the MariaDB table and deleted them. I am curious if it has to do with when the user creates a folder and the sync and/or background sync are not automatically checked.

**Comment by MarZi235** [comment]
Since this issue is blocking production environments - is there any ETA on a fix delivery? We've been experiencing this issue for over a year now - ever since #8223 or #8555. Would you consider prioritizing this, please?

**Comment by ChristophWurst** [comment]
We are working on this. See https://github.com/nextcloud/mail/pull/9649 for example.

**Comment by blacklight** [comment]
I have migrated to SnappyMail in the meantime. Nextcloud Mail is too slow and doesn't even allow bulk operations. This cache issue is definitely the biggest problem it has in my opinion (and it's been open for too long), but it's not the only one plaguing this product. I'll come back and check out Nextcloud Mail when it's more mature.

**Comment by alexh-swdev** [comment]
Same here. SnappyMail is not perfekt, either. But the basic functionality, at least, is working.

**Comment by frederikb96** [comment]
I installed SnappyMail app integrated into Nextcloud, as well. But not sure if I will continue using it. Basic functionality is working and way better, indeed. However, setup of multiple accounts is more cumbersome and no "All inboxes" option, etc.

The idea and concept of Nextcloud Mail is really cool, but I also have too many problems with it currently, and probably will use SnappyMail in the meantime.

**Comment by Cijay22** [comment]
Is there any serious solution for the issue with the ghost emails? Currently, I need to delete the cache for users daily via the debug mode workaround. However, it is also frustrating for the users when they have to contact me repeatedly. Perhaps as a quick solution, add a command line option that we can call via a cron job daily to clean up the cache. But with this issue, the app is unusable. Or what will be happening when I empty the database table nightly is the background sync job filling up the mail boxed again in the background or first when the user is accessing there mailbox. 

It's really unfortunate because the mail app is best integrated into Nextcloud. However, it seems underdeveloped, which is truly incomprehensible when you compare it to Roundcube or Snappy Mail, where such problems have never occurred.

Maybee it helps also for other user to explain the issue from technical point of view waht is happening in this case?!

Thanks!

**Comment by ChristophWurst** [comment]
Update to v3.7.2, reset the cache one more time and let us know if the inconsistency still happens.

**Comment by Chartman123** [comment]
@ChristophWurst I did a quick test on two of our users with 3.7.2 and until now no problems any more... Is there a way to do the cache cleaning automatically via occ or a background job?

**Comment by rikmeijer** [comment]
It looks like it working for me.

**Comment by Chartman123** [comment]
> Are you saying the newly deleted emails are also correctly removed from the inbox without needimg to manually clear the cache again? 

Yes, I did just what Christoph posted... Update the app, enable debugging in NC, clicked the clear cache button of my inbox, deleted a mail in another client, mail was gone in NC mail.

> If so, could you outline the steps you took to upgrade? Did you disable debug mode, reboot you server, restart nextcloud, or any additional steps? I cant seem to get the update to work like you are seeing it work.

I did the update to 3.7.2 manually (download file from releases as my updater didn't show it yet, remove old mail-app folder, unzip the archive and run the update via web). No reboot/restart whatsoever

**Comment by frederikb96** [comment]
Thanks a lot for the update :smiley: 

The new version is also available in the store now :+1: 

I updated, enabled debug mode, cleared cache and disabled debug mode again. After that, I got a mail, deleted it on my phone and the mail was also gone in Nextcloud Mail. Looks like there is a good chance that the issue is fixed now, thanks a lot :partying_face:

**Comment by st3iny** [comment]
> @ChristophWurst I did a quick test on two of our users with 3.7.2 and until now no problems any more... Is there a way to do the cache cleaning automatically via occ or a background job?
> 
> Other than @meichthys I deleted a message for testing... (and with "delete" I mean move mail to trash folder)

No, there is no way to automate or script the clean up for now. We plan to implement a periodic full sync to fully reconcile cached messages once we get confirmation that we fixed the underlying issue.

**Comment by flortsch** [comment]
Updated to 3.7.2, deleted all DB entries in all oc_mail_* tables and re-provisioned all mail accounts. Will test this in the next few days. But it is looking good so far, I moved & deleted a few mails with a different mail client and after hitting the refresh button in Nextcloud mail the deleted / moved mails were also correctly deleted / moved. Some of the mail movements were detected automatically by the app without hitting the refresh button.

**Comment by violoncelloCH** [comment]
After testing this a few times over the last days, I can also confirm this to be fixed in v3.7.2 for the previously affected mailbox.

**Comment by cwerstuik** [comment]
I also have found it's been working well now, but I came across one scenario where it didn't work.

It seems I deleted an email from an external email client, but then I received a 2 more emails in that thread. In Nextcloud mail, all 3 emails showed up in the thread and email 1 showed the "not found" warning and I couldn't delete the whole thread

When I deleted the emails 2 and 3 individually in Nextcloud, the first email also disappeared properly. So technically there's an easy work around, but it's still a glitch and can be confusing

**Comment by untuonac** [comment]
After updating to version 3.7.2, the problem would seem to be solved for me. Thank you all.

**Comment by MarZi235** [comment]
Unfortunately, the issue is not resolved for me. After upgrading to 3.7.2 (running NC 29.0.3.4) and clearing all the caches for all users (via enabling debug and manually clearing the caches), I still see stale e-mail in my inboxes. Anything I can do to help root cause the issue?

**Comment by rikmeijer** [comment]
Yes, I am afraid I still see ghost mails in my inbox :'(

**Comment by powerpaul17** [comment]
Yes, unfortunately I'm still seeing the "ghost" messages. It seems it mostly affects the important ones.

**Comment by violoncelloCH** [comment]
Did you all clean the mailbox cache after updating to v3.7.2 ? Without that the old ghost mails will stay, the fix only affects new mails.

**Comment by alexh-swdev** [comment]
> Did you all clean the mailbox cache after updating to v3.7.2 ? Without that the old ghost mails will stay, the fix only affects new mails.

Isn't that sth the upgrade should do on its own?

**Comment by frederikb96** [comment]
> Thanks a lot for the update 😃
> 
> The new version is also available in the store now 👍
> 
> I updated, enabled debug mode, cleared cache and disabled debug mode again. After that, I got a mail, deleted it on my phone and the mail was also gone in Nextcloud Mail. Looks like there is a good chance that the issue is fixed now, thanks a lot 🥳


After a few days of testing, I sadly have to say that the **issue still exists**. What I did:
- Having a mail account (from posteo.de) being set up in Nextcloud Mail.
- Going on vacation, so not opening Nextcloud Mail or any other Mail client for some time
- Opening the Mail client K9-Mail on my phone: deleting some mails I received during my days off
- Opening Nextcloud Mail, and the once I already deleted via K9-Mail show up as those ghost mails again, which cannot be deleted

Before my vacation, it was fine for two days or so. Probably because, I opened Nextcloud Mail regularly. So deleting mails via other mail client did not have negative impact.

**Comment by philiprenich** [comment]
I'll also throw in that I am still experiencing this issue. I am on v3.7.2 and have put NC into debug mode and used the revealed clear cache function in Mail. This issue unfortunately isn't an uncommon occurrence, it is consistently happening. All externally deleted mail stays in NC Mail. It unfortunately make the app unusable. At this point I am only keeping it installed so I can keep testing new versions and maybe helping with bug reports. To that end, if there are logs I can post, do let me know.

**Comment by pietricello** [comment]
I have the same issue now, with NC 30.0.0 . Differently from other users i was not using mail-app at all due this issue. Yesterday i wanted give another chance but as i configured my imap accounts issue came back. Same as above: when i delete an email it remains visible in inbox but clicking on it return a "not found" error because obviously mail is not on server anymore. Just to say. Looking at this thread is since January people struggling with this and after 9 months and several minor/major release it is still present. Another reason to continue using SnappyMail instead.

**Comment by ChristophWurst** [comment]
https://github.com/nextcloud/mail/pull/10036 will fix the edge case where IMAP pushes vanishes messages

**Comment by LostinSpacetime** [comment]
> #10036 will fix the edge case where IMAP pushes vanishes messages

Will this fix also clean up the inbox on all accounts?

**Comment by ChristophWurst** [comment]
Yes https://github.com/nextcloud/mail/pull/10038. Stay tuned.

**Comment by LostinSpacetime** [comment]
My menu looks the same:
![image](https://github.com/user-attachments/assets/1d292656-1569-4a91-a3e4-3b8901a778b1)
Do I need to do anything to see the option to "Repair Mailbox"?

**Comment by ChristophWurst** [comment]
Wait for the change to be released and update the app.

**Comment by LostinSpacetime** [comment]
I am already running Mail 4.0.1. I thought the change was included in Mail 4.0.1.
